### PR TITLE
Adding ability to expand and columns slickgrid table rows

### PR DESCRIPTION
**Pull request by aasimkhan30**
Collapsed: (Default state)
image
Expanded:
image
Nested:
image

Dark Mode:
image

This PR fixes: #19127

### CODE REVIEW

File: src/sql/base/browser/ui/table/formatters.ts [hunk 1]
```diff
@@ -166,6 +166,42 @@ export function slickGridDataItemColumnValueWithNoData(value: any, columnDef: an
...
+		} else {
+			valueToDisplay = value;
+		}
+		valueToDisplay = escape(valueToDisplay.length > 250 ? valueToDisplay.slice(0, 250) + '...' : valueToDisplay);
+		titleValue = valueToDisplay;
+	}
+
+	if (dataContext['isParent']) {
+		if (dataContext._collapsed) {
+			return spacer + `<span class='codicon codicon-chevron-down toggle'></span>&nbsp; <span title="${titleValue}" class="${cellClasses}">${valueToDisplay}</span>`;
```
Comment (Charles-Gagnon) [inline src/sql/base/browser/ui/table/formatters.ts:196]: Is this accessible? (screen reader, keyboard navigation and making sure it's announcing the number of rows correctly)

(I would suggest going to the a11y office hours and asking them about how to make a table with collapsible rows like this accessible)

File: src/sql/base/browser/ui/table/tableDataView.ts [hunk 2]
```diff
@@ -63,6 +64,19 @@ function defaultFilter<T extends Slick.SlickData>(data: T[], columns: Filterable
 			});
 		}
 	});
+	const collapsibleTable = columns.find(c => c.formatter === expandableColumnFormatter);
+	if (collapsibleTable) {
+		filteredData = filteredData.filter((item) => {
+			let parent = data[item.parent];
+			while (parent) {
+				if (!parent._collapsed) {
+					return false;
```
Comment (Charles-Gagnon) [inline src/sql/base/browser/ui/table/tableDataView.ts:73]: I'm not sure I understand this - if I'm reading this correctly isn't this filtering out the child elements from the data who have parents that are collapsed? But then shouldn't this be returning false if it finds a parent that's collapsed and default to true otherwise?

File: src/sql/base/browser/ui/table/table.ts [hunk 3]
```diff
@@ -121,6 +121,38 @@ export class Table<T extends Slick.SlickData> extends Widget implements IDisposa
...
+					(<any>row)._collapsed = false;
+				}
+				data.grid.updateRow(data.row);
+				this._data.filter(data.grid.getColumns());
+				this.rerenderGrid();
+			}
+		});
+
+		this._grid.onKeyDown.subscribe((e, data) => {
+			if ((<any>e).keyCode === 13) {
```
Comment (Charles-Gagnon) [inline src/sql/base/browser/ui/table/table.ts:140]: Space as well

**Review by Charles-Gagnon — CHANGES_REQUESTED** [review]

File: src/sql/base/browser/ui/table/formatters.ts (hunk 1 above)
Comment (erinstellato-ms) [reply src/sql/base/browser/ui/table/formatters.ts:196]: Definitely support checking on accessibility here and getting feedback if needed.

**Review by erinstellato-ms — APPROVED** [review]
Visually this looks good to me, as long as it is meets accessibility standards.  The indent helps ensure folks notice that line is different (when collapsed), I don't know if the chevron needs to be darker/bold to help make it noticeable.

File: src/sql/base/browser/ui/table/tableDataView.ts (hunk 2 above)
Comment (aasimkhan30) [reply src/sql/base/browser/ui/table/tableDataView.ts:73]: Fixed the variable name. This code will make more sense.

File: src/sql/base/browser/ui/table/table.ts (hunk 3 above)
Comment (aasimkhan30) [reply src/sql/base/browser/ui/table/table.ts:140]: Space is also tied to scrolling down of the grid. Should I stop propagation?

Comment (Charles-Gagnon) [reply src/sql/base/browser/ui/table/table.ts:140]: Space scrolls the grid? That's....weird. I don't think that's normal behavior. @alanrenmsft ?

File: src/sql/base/browser/ui/table/formatters.ts [hunk 4]
```diff
@@ -166,6 +166,42 @@ export function slickGridDataItemColumnValueWithNoData(value: any, columnDef: an
 	};
 }
 
+export function expandableColumnFormatter(row: number | undefined, cell: any | undefined, value: any, columnDef: any | undefined, dataContext: any | undefined): string {
+	const spacer = `<span style='display:inline-block;height:1px;width:" + (15 * dataContext["indent"]) + "px'></span>`;
+	let cellClasses = 'grid-cell-value-container';
+	let valueToDisplay = '';
+	let titleValue = '';
+
+	if (DBCellValue.isDBCellValue(value)) {
```
[the remark this answers is not in the document]
Comment (aasimkhan30) [reply src/sql/base/browser/ui/table/formatters.ts:175]: Yes, I got the logic from textFormatter. Will consolidate it to a helper function

**Comment by aasimkhan30** [comment]
@erinstellato-ms, updated screenshots with the higher contrast icon.

**Comment by alanrenmsft** [comment]
my main concern is accessibility, I don't remember seeing any accessibility defects opened for SSMS for this feature, could you please attend the accessibility office hour and understand the accessibility requirements for features like this?

File: src/sql/base/browser/ui/table/table.ts [hunk 5]
```diff
@@ -405,4 +463,59 @@ export class Table<T extends Slick.SlickData> extends Widget implements IDisposa
...
+	private isTreeGrid(): boolean {
+		return this._grid.getColumns().filter(c => c.formatter === treeGridExpandableColumnFormatter).length > 0;
+	}
+
+	// Gets the index for the expandable column
+	private expandableColumnIndex(): number {
+		return this._grid.getColumns().findIndex(c => c.formatter === treeGridExpandableColumnFormatter);
+	}
+
+	// This expands/collapse the tree grid item
```
Comment (Charles-Gagnon) [inline src/sql/base/browser/ui/table/table.ts:477]: Use actual JS docs for all these (with params)

File: src/sql/base/browser/ui/table/table.ts [hunk 6]
```diff
@@ -405,4 +463,59 @@ export class Table<T extends Slick.SlickData> extends Widget implements IDisposa
...
+			this._data.filter(this._grid.getColumns());
+			this.rerenderGrid();
+			this.focus();
+		}
+	}
+
+	// We need to transform grid data to tree grid data. This includes adding required aria attributes to the rows
+	private transGridDataToTreeGridData(data: IDisposableDataProvider<T>): IDisposableDataProvider<T> {
+		for (let i = 0; i < data.getLength(); i++) {
+			const dataRow = <any>data.getItem(i);
```
Comment (Charles-Gagnon) [inline src/sql/base/browser/ui/table/table.ts:499]: This should be T - why are you casting it to any? If the type isn't right we should be fixing that instead

File: src/sql/base/browser/ui/table/table.ts [hunk 7]
```diff
@@ -405,4 +463,59 @@ export class Table<T extends Slick.SlickData> extends Widget implements IDisposa
...
+				(<any>rowData).expanded = forceState;
+			}
+			this._data.filter(this._grid.getColumns());
+			this.rerenderGrid();
+			this.focus();
+		}
+	}
+
+	// We need to transform grid data to tree grid data. This includes adding required aria attributes to the rows
+	private transGridDataToTreeGridData(data: IDisposableDataProvider<T>): IDisposableDataProvider<T> {
```
Comment (Charles-Gagnon) [inline src/sql/base/browser/ui/table/table.ts:497]: You should avoid using shortened terms (especially ones like trans that have other potentially sensitive meanings)

File: src/sql/base/browser/ui/table/tableDataView.ts [hunk 8]
```diff
@@ -63,6 +64,20 @@ function defaultFilter<T extends Slick.SlickData>(data: T[], columns: Filterable
 			});
 		}
 	});
+
+	const isTreeGrid = columns.find(c => c.formatter === treeGridExpandableColumnFormatter);
+	if (isTreeGrid) {
+		filteredData = filteredData.filter((item) => {
```
Comment (Charles-Gagnon) [inline src/sql/base/browser/ui/table/tableDataView.ts:70]: Add comment here explaining what this is supposed to be doing

File: src/sql/workbench/contrib/executionPlan/browser/executionPlanPropertiesView.ts [hunk 9]
```diff
@@ -131,8 +131,9 @@ export class ExecutionPlanPropertiesView extends ExecutionPlanPropertiesViewBase
 		props.forEach((p, i) => {
 			let row = {};
 			rows.push(row);
-			row['name'] = '  '.repeat(indent) + p.name;
+			row['name'] = p.name;
 			row['parent'] = parentIndex;
+			//row['indent'] = indent;
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/executionPlan/browser/executionPlanPropertiesView.ts:136]: remove

File: src/sql/base/browser/ui/table/table.ts [hunk 10]
```diff
@@ -405,4 +463,59 @@ export class Table<T extends Slick.SlickData> extends Widget implements IDisposa
 	public get container(): HTMLElement {
 		return this._tableContainer;
 	}
+
+	// Checks if the current table is a tree grid. Every tree grid should have a column with expandableColumnFormatter
+	private isTreeGrid(): boolean {
```
Comment (Charles-Gagnon) [inline src/sql/base/browser/ui/table/table.ts:468]: Instead of having all this custom logic for a specific type of "grid" (tree grid in this case) I think it would be a lot better to create a new class that extends Table and modifies the behavior as needed. This will be a lot easier to work with and reduce the overhead you're putting on other tables with all this added logic.

**Review by Charles-Gagnon — CHANGES_REQUESTED** [review]

**Comment by alanrenmsft** [comment]
@aasimkhan30 I guess this is not really ready for review since you have to make changes in slickgrid and then vbump it here, right?

**Comment by aasimkhan30** [comment]
@alanrenmsft, I am using slickgrid's onRendered event to add the necessary attributes. This does not require any chagnes to the slickgrid libarary.

File: src/sql/base/browser/ui/table/table.ts [hunk 11]
```diff
@@ -121,6 +123,58 @@ export class Table<T extends Slick.SlickData> extends Widget implements IDisposa
...
+
+		this._grid.onClick.subscribe((e, data) => {
+			if (this.isTreeGrid()) {
+				this.toggleTreeGridParent(data.row, data.cell);
+			}
+		});
+
+		this._grid.onKeyDown.subscribe((e, data) => {
+			if (this.isTreeGrid()) {
+				if ((<any>e).keyCode === 13) {
```
Comment (alanrenmsft) [inline src/sql/base/browser/ui/table/table.ts:135]: please look for other places that we handle keyboard event and use the keycode name instead of number.

File: src/sql/base/browser/ui/table/table.ts [hunk 12]
```diff
@@ -121,6 +123,58 @@ export class Table<T extends Slick.SlickData> extends Widget implements IDisposa
...
 		this._grid.onColumnsResized.subscribe(() => this._onColumnResize.fire());
+
+		this._grid.onClick.subscribe((e, data) => {
+			if (this.isTreeGrid()) {
+				this.toggleTreeGridParent(data.row, data.cell);
+			}
+		});
+
+		this._grid.onKeyDown.subscribe((e, data) => {
+			if (this.isTreeGrid()) {
```
Comment (alanrenmsft) [inline src/sql/base/browser/ui/table/table.ts:134]: you also need to mark the event as handled and stop propagation , right?

File: src/sql/base/browser/ui/table/table.ts [hunk 13]
```diff
@@ -121,6 +123,58 @@ export class Table<T extends Slick.SlickData> extends Widget implements IDisposa
...
+					}
+				}
+			}
+		});
+
+		this._grid.onRendered.subscribe((e, data) => {
+			if (this.isTreeGrid()) {
+				// Changing table role from grid to treegrid
+				this._tableContainer.setAttribute('role', 'treegrid');
+				for (let i = 0; i < this._data.getLength(); i++) {
```
Comment (alanrenmsft) [inline src/sql/base/browser/ui/table/table.ts:156]: this is not efficient; you should only process the visible rows:  use this.grid.getViewPort() method to get the view port, use the top and bottom properties as start/end rows.

File: src/sql/base/browser/ui/table/table.ts [hunk 14]
```diff
@@ -121,6 +123,58 @@ export class Table<T extends Slick.SlickData> extends Widget implements IDisposa
...
+					const rowData = this._data.getItem(i);
+					// Getting the row div that corresponds to the data row
+					const rowElement = this._tableContainer.querySelector(`div [role="row"][aria-rowindex="${(i + 1)}"]`);
+					// If the row element is found in the dom, we are setting the required aria attributes for it.
+					if (rowElement) {
```
Comment (alanrenmsft) [inline src/sql/base/browser/ui/table/table.ts:161]: is the row html element not being reused by slickgrid? if it is you also need to clear the values you set previously for a different row.

File: src/sql/base/browser/ui/table/table.ts (hunk 10 above)
Comment (alanrenmsft) [reply src/sql/base/browser/ui/table/table.ts:468]: yes, I wasn't expecting this many changes in the file, now I also think you should have a subclass for it.

File: src/sql/base/browser/ui/table/table.ts [hunk 15]
```diff
@@ -405,4 +463,59 @@ export class Table<T extends Slick.SlickData> extends Widget implements IDisposa
 	public get container(): HTMLElement {
 		return this._tableContainer;
 	}
+
+	// Checks if the current table is a tree grid. Every tree grid should have a column with expandableColumnFormatter
+	private isTreeGrid(): boolean {
+		return this._grid.getColumns().filter(c => c.formatter === treeGridExpandableColumnFormatter).length > 0;
```
Comment (alanrenmsft) [inline src/sql/base/browser/ui/table/table.ts:469]: looks like you should introduce a property in the column definition, this doesn't seem to be a proper way to achieve what you want.

File: src/sql/base/browser/ui/table/table.ts [hunk 16]
```diff
@@ -405,4 +463,59 @@ export class Table<T extends Slick.SlickData> extends Widget implements IDisposa
...
+	}
+
+	// Gets the index for the expandable column
+	private expandableColumnIndex(): number {
+		return this._grid.getColumns().findIndex(c => c.formatter === treeGridExpandableColumnFormatter);
```
Comment (alanrenmsft) [inline src/sql/base/browser/ui/table/table.ts:474]: what does expandable column mean? in the changes above, I saw that you have code to expand/collapse a row when precess left/right keys in the first/last cell.

File: src/sql/base/browser/ui/table/table.ts (hunk 7 above)
Comment (alanrenmsft) [reply src/sql/base/browser/ui/table/table.ts:497]: it does bother me too :)

File: src/sql/base/browser/ui/table/table.ts [hunk 17]
```diff
@@ -121,6 +123,58 @@ export class Table<T extends Slick.SlickData> extends Widget implements IDisposa
 		this.mapMouseEvent(this._grid.onHeaderClick, this._onHeaderClick);
 		this.mapMouseEvent(this._grid.onDblClick, this._onDoubleClick);
 		this._grid.onColumnsResized.subscribe(() => this._onColumnResize.fire());
+
```
Comment (alanrenmsft) [inline src/sql/base/browser/ui/table/table.ts:126]: test code for tree grid?

File: src/sql/base/browser/ui/table/table.ts (hunk 16 above)
Comment (aasimkhan30) [reply src/sql/base/browser/ui/table/table.ts:474]: It is the column which contains the chevron icon 

image

File: src/sql/base/browser/ui/table/table.ts (hunk 15 above)
Comment (aasimkhan30) [reply src/sql/base/browser/ui/table/table.ts]: This is not needed as I have made a new TreeGrid component.

File: src/sql/base/browser/ui/table/table.ts (hunk 14 above)
Comment (aasimkhan30) [reply src/sql/base/browser/ui/table/table.ts]: It does not really matter because I am changing the attributes for all the visible rows. So, even if they are reused, my for loop should take care of it.

File: src/sql/base/browser/ui/table/table.ts [hunk 18]
```diff
@@ -405,4 +406,7 @@ export class Table<T extends Slick.SlickData> extends Widget implements IDisposa
 	public get container(): HTMLElement {
 		return this._tableContainer;
 	}
+
```
Comment (alanrenmsft) [inline src/sql/base/browser/ui/table/table.ts:409]: please remove

File: src/sql/base/browser/ui/table/treeGrid.ts [hunk 19]
```diff
@@ -0,0 +1,176 @@
...
+						this.setCellExpandedState(data.row, this.expandableColumnIndex(), false);
+					}
+				} else if (event.keyCode === KeyCode.RightArrow) {
+					// Right arrow on last cell of the collapsed row expands it.
+					if (data.cell === (this._grid.getColumns().length - 1)) {
+						this.setCellExpandedState(data.row, this.expandableColumnIndex(), true);
+					}
+				}
+			}
+			return false;
```
Comment (alanrenmsft) [inline src/sql/base/browser/ui/table/treeGrid.ts:77]: what is this for?

Comment (aasimkhan30) [reply src/sql/base/browser/ui/table/treeGrid.ts:77]: Since these are Jquery events, returning false does preventDefault and stopPropogation.

**Review by erinstellato-ms — COMMENTED** [review]
No additional feedback at this time (I only see code changes but let me know if I missed something visual to review).

File: src/sql/base/browser/ui/table/table.ts (hunk 14 above)
Comment (alanrenmsft) [reply src/sql/base/browser/ui/table/table.ts]: but you do have null/undefined check, so it is possible that you are not clearing the attribute that were set previously.

File: src/sql/base/browser/ui/table/treeGrid.ts [hunk 20]
```diff
@@ -0,0 +1,177 @@
...
+
+
+export class TreeGrid<T extends Slick.SlickData> extends Table<T> {
+	constructor(parent: HTMLElement, configuration?: ITableConfiguration<T>, options?: Slick.GridOptions<T>) {
+		super(parent, configuration, options);
+
+		// Changing table role from grid to treegrid
+		this._tableContainer.setAttribute('role', 'treegrid');
+
+		if (!configuration || !configuration.dataProvider || isArray(configuration.dataProvider)) {
```
Comment (alanrenmsft) [inline src/sql/base/browser/ui/table/treeGrid.ts:46]: if you invert the logic the check can be simplified.

File: src/sql/base/browser/ui/table/treeGrid.ts [hunk 21]
```diff
@@ -0,0 +1,177 @@
...
+				(<any>rowData).expanded = expanded;
+			}
+			this._data.filter(this._grid.getColumns());
+			this.rerenderGrid();
+			this.focus();
+		}
+	}
+
+	/**
+	 * Gets the index for the expandable column
```
Comment (alanrenmsft) [inline src/sql/base/browser/ui/table/treeGrid.ts:144]: please add description what does this mean.

File: src/sql/base/browser/ui/table/table.ts (hunk 14 above)
Comment (alanrenmsft) [reply src/sql/base/browser/ui/table/table.ts]: nvm, it looks like you are setting the values for every row in transform. so that should be good.

File: src/sql/base/browser/ui/table/treeGrid.ts [hunk 22]
```diff
@@ -0,0 +1,177 @@
...
+			this.focus();
+		}
+	}
+
+	/**
+	 * Gets the index of the column that has expandable column formatter. This column contains the chevron
+	 * icon that indicates if the row is expanded or collapsed.
+	 */
+	private getExpandableColumnIndex(): number {
+		return this._grid.getColumns().findIndex(c => c.formatter === treeGridExpandableColumnFormatter);
```
Comment (Charles-Gagnon) [inline src/sql/base/browser/ui/table/treeGrid.ts:147]: Something like this would always be the first column, right? Or at least I'm not sure of a scenario where it wouldn't be that - so for ease of implementation I'd suggest just making this only support having the first column. 

And overall I'm not sure I like the pattern of having the formatter for that column need to be defined by the user directly and always have to be treeGridExpandableColumnFormatter. This is a tree grid so we should be assuming it's always going to have this formatter. So that should be something that this class handles - not something the user does. 

I'd suggest having the user specify a column formatter as normal and then in this class when the columns are set you create a treeGridExpandableColumnFormatter that has a parameter for the "base" formatter (defaulting to text cell if none specified) and overwrite the first column to have this new formatter. Then when the formatter is called it calls the base formatter to get the "inner" content and wraps that in the div like you have above. (this is essentially the decorator pattern)

File: src/sql/base/browser/ui/table/treeGrid.ts [hunk 23]
```diff
@@ -0,0 +1,177 @@
...
+	 * This functions toggles the expanded state of a parent cell.
+	 * @param row row index of the parent cell
+	 * @param cell cell/ column index of the parent cell
+	 * @param expanded force a state on the parent cell
+	 */
+	private setCellExpandedState(row: number, cell: number, expanded?: boolean): void {
+		const rowData = this._data.getItem(row);
+		if (rowData['isParent'] && this._grid.getColumns()[cell].formatter === treeGridExpandableColumnFormatter) {
+			if (expanded === undefined) {
+				(<any>rowData).expanded = !rowData.expanded;
```
Comment (aasimkhan30) [inline src/sql/base/browser/ui/table/treeGrid.ts:132]: @Charles-Gagnon this particular line toggles the expanded state, doesn't it?

File: src/sql/base/browser/ui/table/treeGrid.ts (hunk 22 above)
Comment (Charles-Gagnon) [reply src/sql/base/browser/ui/table/treeGrid.ts:147]: Where?

Honestly though I'd try to avoid over-engineering this and just make it as simple as we can right now. Always having the first column be the expandable one is a lot simpler and lets us focus on making sure that everything works in that scenario and then later on if needed we can modify it to support other columns. 

(for example - if you have other columns then you have to consider what if someone sets expandable to true on multiple columns?)

File: src/sql/base/browser/ui/table/treeGrid.ts [hunk 24]
```diff
@@ -0,0 +1,167 @@
...
+		}
+		this.addTreeGridDataAttributes(this._data);
+		this._grid.setData(this._data, true);
+		this._data.filter(this._grid.getColumns());
+	}
+
+	/**
+	 *  Sets the expanded state to the specified value, or if undefined toggles the current state of the cell
+	 */
+	private expandRow(row: number, expanded?: boolean): void {
```
Comment (Charles-Gagnon) [inline src/sql/base/browser/ui/table/treeGrid.ts:127]: This still isn't a very good name - it can also collapse the row. Was there something wrong with `setRowExpandedState`?

File: src/sql/base/browser/ui/table/treeGrid.ts [hunk 25]
```diff
@@ -0,0 +1,167 @@
...
+				(<any>rowData).expanded = expanded;
+			}
+			this._data.filter(this._grid.getColumns());
+			this.rerenderGrid();
+			this.focus();
+		}
+	}
+
+	/**
+	 * We need to augment the grid data to include additional properties that are necessary for rendering a tree grid.
```
Comment (Charles-Gagnon) [inline src/sql/base/browser/ui/table/treeGrid.ts:142]: Typically comments don't talk in first person, you just need to describe what it's doing 

`Adds additional properties to data rows necessary for displaying as part of the tree grid structure`

File: src/sql/base/browser/ui/table/media/table.css [hunk 26]
```diff
@@ -234,3 +234,8 @@
 	display: flex;
 	align-items: center;
 }
+
+.slick-cell .toggle {
+	font-size: 12px !important;
```
Comment (alanrenmsft) [inline src/sql/base/browser/ui/table/media/table.css:239]: better to understand why you have to use important.

File: src/sql/base/browser/ui/table/treeGrid.ts [hunk 27]
```diff
@@ -0,0 +1,167 @@
...
+						return false;
+					}
+				}
+			}
+			return true;
+		});
+
+		this._grid.onRendered.subscribe((e, data) => {
+			const visibleRows = this._grid.getViewport();
+			for (let i = visibleRows.top; i < visibleRows.bottom; i++) {
```
Comment (alanrenmsft) [inline src/sql/base/browser/ui/table/treeGrid.ts:86]: i<=visibleRow.bottom

I think it should be inclusive

File: src/sql/base/browser/ui/table/treeGrid.ts [hunk 28]
```diff
@@ -0,0 +1,167 @@
...
+				const rowData = this._data.getItem(i);
+				// Getting the row div that corresponds to the data row
+				const rowElement = this._tableContainer.querySelector(`div [role="row"][aria-rowindex="${(i + 1)}"]`);
+				// If the row element is found in the dom, we are setting the required aria attributes for it.
+				if (rowElement) {
+					if (rowData.expanded !== undefined) {
```
Comment (alanrenmsft) [inline src/sql/base/browser/ui/table/treeGrid.ts:92]: have you confirmed that slickgrid is not reusing the row container? if it does reuse, then not all rowdata has expanded set, and there will be a problem.

File: src/sql/workbench/contrib/executionPlan/browser/executionPlanPropertiesView.ts [hunk 29]
```diff
@@ -143,10 +143,11 @@ export class ExecutionPlanPropertiesView extends ExecutionPlanPropertiesViewBase
 					text: removeLineBreaks(p.displayValue, ' '),
 					style: parentRowCellStyling
 				};
+				row['tootltip'] = p.displayValue;
 				this.convertModelToTableRows(p.value, rows.length - 1, indent + 2, rows);
 			} else {
-				row['value'] = removeLineBreaks(p.value, ' ');
-				row['tooltip'] = p.value;
+				row['value'] = removeLineBreaks(p.displayValue, ' ');
```
Comment (alanrenmsft) [inline src/sql/workbench/contrib/executionPlan/browser/executionPlanPropertiesView.ts:149]: name, parent, value, tooltip, you should define an enum for these values.

**Review by alanrenmsft — APPROVED** [review]
overall looks good to me, I left a couple more comments, please address them.

File: src/sql/base/browser/ui/table/treeGrid.ts (hunk 28 above)
Comment (aasimkhan30) [reply src/sql/base/browser/ui/table/treeGrid.ts:92]: https://github.com/microsoft/SlickGrid.ADS/blob/08f1ca3f6eb2cd1ea7f56e2af4b69600d48e1932/slick.grid.js#L2390

Slickgrid used a different element for every row. It might store those elements in the cache, but it will still only reuse it for the same row.

Comment (aasimkhan30) [reply src/sql/base/browser/ui/table/treeGrid.ts:92]: I will still unset the attributes if the properties are undefined. In that way we should not see any such issues.

File: src/sql/base/browser/ui/table/formatters.ts [hunk 30]
```diff
@@ -170,6 +143,55 @@ export function slickGridDataItemColumnValueWithNoData(value: any, columnDef: an
 	};
 }
 
+export function treeGridExpandableColumnFormatter<T>(formattingFunction: Slick.Formatter<T>) {
```
Comment (Charles-Gagnon) [inline src/sql/base/browser/ui/table/formatters.ts:146]: 1. Add a doc comment explaining what this one is for and why it's different than the others
2. Would suggest `createTreeGridExpandableColumnFormatter` as the name - this function isn't the formatter itself
3. Add return type

File: src/sql/base/browser/ui/table/treeGrid.ts [hunk 31]
```diff
@@ -0,0 +1,183 @@
...
+
+/**
+ * TreeGrid component displays a hierarchical table data grouped into expandable and collapsible nodes.
+ */
+export class TreeGrid<T extends Slick.SlickData> extends Table<T> {
+	constructor(parent: HTMLElement, configuration?: ITableConfiguration<T>, options?: Slick.GridOptions<T>) {
+		super(parent, configuration, options);
+		this._tableContainer.setAttribute('role', 'treegrid');
+		if (this._columns[0]) {
+			this._columns[0].formatter = this._columns[0].formatter ? treeGridExpandableColumnFormatter(this._columns[0].formatter) : treeGridExpandableColumnFormatter(textFormatter);
```
Comment (Charles-Gagnon) [inline src/sql/base/browser/ui/table/treeGrid.ts:44]: `this._columns[0].formatter = treeGridExpandableColumnFormatter(this._columns[0].formatter ?? textFormatter) ;`

File: src/sql/base/browser/ui/table/treeGrid.ts [hunk 32]
```diff
@@ -0,0 +1,183 @@
...
+}
+
+/**
+ * TreeGrid component displays a hierarchical table data grouped into expandable and collapsible nodes.
+ */
+export class TreeGrid<T extends Slick.SlickData> extends Table<T> {
+	constructor(parent: HTMLElement, configuration?: ITableConfiguration<T>, options?: Slick.GridOptions<T>) {
+		super(parent, configuration, options);
+		this._tableContainer.setAttribute('role', 'treegrid');
+		if (this._columns[0]) {
```
Comment (Charles-Gagnon) [inline src/sql/base/browser/ui/table/treeGrid.ts:43]: Add comment explaining what this is doing

File: src/sql/base/browser/ui/table/treeGrid.ts [hunk 33]
```diff
@@ -0,0 +1,183 @@
...
+		} else {
+			this._data = new TableDataView<T>(data, undefined, undefined, defaultTreeGridFilter);
+		}
+		this.addTreeGridDataAttributes(this._data);
+		this._grid.setData(this._data, true);
+		this._data.filter(this._grid.getColumns());
+	}
+
+	public override set columns(columns: Slick.Column<T>[]) {
+		if (columns[0]) {
```
Comment (Charles-Gagnon) [inline src/sql/base/browser/ui/table/treeGrid.ts:134]: This is duplicated with the code in the constructor. Consolidate this into one - preferably by updating the base constructor to call this property setter instead.

File: src/sql/workbench/contrib/executionPlan/browser/executionPlanPropertiesView.ts [hunk 34]
```diff
@@ -93,7 +93,7 @@ export class ExecutionPlanPropertiesView extends ExecutionPlanPropertiesViewBase
 				width: 250,
 				editor: Slick.Editors.Text,
 				headerCssClass: 'prop-table-header',
-				formatter: textFormatter
+				formatter: treeGridExpandableColumnFormatter
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/executionPlan/browser/executionPlanPropertiesView.ts:96]: Revert this change, it should just be textFormatter now

File: src/sql/base/browser/ui/table/formatters.ts [hunk 35]
```diff
@@ -170,6 +143,55 @@ export function slickGridDataItemColumnValueWithNoData(value: any, columnDef: an
...
+				return `<div>${spacer}<span class='codicon codicon-chevron-right toggle' style='font-weight:bold;'></span>&nbsp; ${innerCellContent}</div>`;
+			}
+		} else {
+			return `${spacer}${innerCellContent}`;
+		}
+	};
+}
+
+
+function createTextCell(value: any): string {
```
Comment (Charles-Gagnon) [inline src/sql/base/browser/ui/table/formatters.ts:165]: Now that we're not using this in the new formatter you shouldn't need to separate this out

**Review by Charles-Gagnon — CHANGES_REQUESTED** [review]
